### PR TITLE
Enable developers to specify tooltip position before users interact

### PR DESCRIPTION
**Pull request by julianna-langston**
## Description

Added a `defaultIndex` prop to <strike>all chart types</strike> EDIT: the `<Tooltip />` element. If someone specifies `defaultIndex={2}`, the 2nd data item will show a tooltip. This includes, when relevant:
- The relevant data will show its hover state
- The vertical line / crosshair appears around the data point
- The floating tooltip label will appear in an appropriate location

Additionally, this works in tandem with the accessibilityManager. If a developer provides the prop `defaultIndex={2}`, and a keyboard-only user tabs to that chart, they will jump to the starting position. If the user presses the right arrow, they will move to index 3.

A story has been added under the storybook "Tooltip" section to demonstrate this functionality

Other behavioral notes:
- Supports vertical charts
- Setting `defaultIndex` doesn't tamper with the tooltip's usual interactions. Users can still move around the chart, and the tooltip will move with them.
- If the `defaultIndex` prop is updated, the chart handles the update as expected. (Hence, changing the `componentDidUpdate` lifecycle method in the code.
- If the chart is in a ResponsiveContainer, and the user hasn't interacted with the chart, and the chart gets resized... then the tooltip stays in the correct location.
- Extra handling for scatter plots, since they're weird. (They require a crosshair, which means that we need to know the y-value of the relevant point, but... we don't have that information in the props or state... so I dig it out of the graphicalItem associated with the scatter plot. Don't love this solution, but it's the only way I could find to support scatter plots for this.)

## Related Issue

https://github.com/recharts/recharts/issues/3962

## Motivation and Context

In my library, Chart2Music, we provide extra accessibility support to data visualizations by turning charts into music. As part of that process, we provide screen reader users the ability to access and interact with data in a chart, and visually synchronize the chart to the user's actions.

While we can provide visual syncing with other charting libraries, such as react-chartjs-2, it hasn't been available for recharts.

With this change, we'll be able to improve the UX for users with a Chart2Music+Recharts tech stack.

## How Has This Been Tested?

Added automated tests cover every line that was changed in this PR.

To identify edge cases, I made sure the property would work as expected on every example in storybook.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ckifer** [comment]
Why should this exist on each chart type and not on `Tooltip`? Also couldn't someone set a default index with `activeIndex`

whats the difference between active index and default index? I'd like to avoid bloat to the API if we can avoid it (the functionality isn't bloat of course, but we don't need two fields that accomplish the same thing)

**Comment by julianna-langston** [comment]
My understanding was that `activeTooltipIndex` was internal. In this PR, `defaultIndex` is exposed to recharts consumers, so that they can set the index they want the tooltip to appear at.

I used the term `defaultIndex`, because I didn't want the tooltip to be locked to the provided index. I was concerned that providing `activeTooltipIndex` as a prop would communicate that there's where the tooltip would be, regardless of user interaction.

**Comment by nikolasrieble** [comment]
I would like to also ask for review from @PavelVanecek because he recently worked on the Tooltip.

**Comment by ckifer** [comment]
@julianna-langston gotcha makes more sense. I'll probably pull it down and check out more. 

Curious about how this and the active prop on Tooltip interact

File: src/chart/generateCategoricalChart.tsx [hunk 1]
```diff
@@ -1103,6 +1104,69 @@ export const generateCategoricalChart = ({
         mouseHandlerCallback: this.triggeredAfterMouseMove,
         layout: this.props.layout,
       });
+      this.displayDefaultTooltip();
+    }
+
+    displayDefaultTooltip() {
+      const { defaultIndex } = this.props;
+      if (
+        typeof this.props.defaultIndex !== 'number' ||
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1113]: any particular reason for the number check? TypeScript should take care of things

File: src/chart/generateCategoricalChart.tsx [hunk 2]
```diff
@@ -1103,6 +1104,69 @@ export const generateCategoricalChart = ({
...
+        defaultIndex < 0 ||
+        defaultIndex > this.state.tooltipTicks.length
+      ) {
+        return;
+      }
+
+      const tooltipElem = (Array.isArray(this.props.children) ? this.props.children : [this.props.children]).find(
+        (child: any) => child.type.name === 'Tooltip',
+      );
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1122]: There is a helper function `findChildByType` that could be used here, but also we can hopefully get rid of this soon and instead use context..

File: src/chart/generateCategoricalChart.tsx [hunk 3]
```diff
@@ -1103,6 +1104,69 @@ export const generateCategoricalChart = ({
...
+        ? {
+            x: independentAxisCoord,
+            y: dependentAxisCoord,
+          }
+        : {
+            y: independentAxisCoord,
+            x: dependentAxisCoord,
+          };
+
+      // If there's a scatter plot, we'll want to grab the
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1145]: nit: comment is cut off

File: src/chart/generateCategoricalChart.tsx [hunk 4]
```diff
@@ -1471,8 +1543,11 @@ export const generateCategoricalChart = ({
       const mouse = this.getMouseInfo(e);
       const nextState: CategoricalChartState = mouse ? { ...mouse, isTooltipActive: true } : { isTooltipActive: false };
 
-      this.setState(nextState);
-      this.triggerSyncEvent(nextState);
+      // If the dev set a defaultIndex, don't set `isTooltipActive` to false
+      if (nextState.isTooltipActive || typeof this.props.defaultIndex !== 'number') {
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1547]: don't think we need the number check here

File: src/chart/generateCategoricalChart.tsx [hunk 5]
```diff
@@ -1499,9 +1574,11 @@ export const generateCategoricalChart = ({
      * @return {Object} no return
      */
     handleItemMouseLeave = () => {
-      this.setState(() => ({
-        isTooltipActive: false,
-      }));
+      if (typeof this.props.defaultIndex === 'number') {
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1577]: why the number checks?

File: src/chart/generateCategoricalChart.tsx [hunk 6]
```diff
@@ -1523,8 +1600,10 @@ export const generateCategoricalChart = ({
     handleMouseLeave = (e: any) => {
       const nextState: CategoricalChartState = { isTooltipActive: false };
 
-      this.setState(nextState);
-      this.triggerSyncEvent(nextState);
+      if (typeof this.props.defaultIndex !== 'number') {
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1603]: same here

**Review by ckifer — CHANGES_REQUESTED** [review]
Some nits

File: src/chart/generateCategoricalChart.tsx [hunk 7]
```diff
@@ -1200,8 +1269,8 @@ export const generateCategoricalChart = ({
 
         const newState = {
           ...defaultState,
-          ...keepFromPrevState,
           ...updatesToState,
+          ...keepFromPrevState,
```
Comment (PavelVanecek) [inline src/chart/generateCategoricalChart.tsx:1273]: Why this change?

File: src/chart/generateCategoricalChart.tsx [hunk 8]
```diff
@@ -839,6 +839,7 @@ export interface CategoricalChartProps {
   accessibilityLayer?: boolean;
   role?: string;
   tabIndex?: number;
+  defaultIndex?: number;
```
Comment (PavelVanecek) [inline src/chart/generateCategoricalChart.tsx:842]: Why is this a property of the chart? Could it be on the Tooltip instead?

**Comment by PavelVanecek** [comment]
Okay code changes I commented in code. Now I am thinking library changes.

First, if the only difference between `defaultIndex` and `activeTooltipIndex` is that activeTooltipIndex is internal and undocumented, then let's fix it by making one external and adding documentation and tests. If we want a different behaviour then let's thinking about a name that better captures the difference.

Second, documentation - I think this can be easily confused and needs code comments, and examples, and tests.

Third, tests again! Let's aim for 100% code coverage. Can we get all the new changes tested? Perhaps I can help with some testing techniques to get everything under control. If it's impossible to have a unit test, then let's add more storybooks

File: src/chart/generateCategoricalChart.tsx (hunk 4 above)
Comment (julianna-langston) [reply src/chart/generateCategoricalChart.tsx]: So, I figured that if someone wanted the tooltip to be visible from the beginning, that they'd always want it to be visible. 
So, if `defaultIndex` was set, I don't want `isTooltipActive` to ever be false.
So, in the 2 `*MouseLeave` methods, I'd only allow `isTooltipActive=false` if `defaultIndex` _wasn't_ set. 

Using `typeof defaultIndex === "number"` was just how I was checking if it had been set.

At this point, I think I was just adding too much complexity. I think the decision to make the tooltips always visible should be separate from where the tooltip should appear by default. I've backed out these checks, to reduce the scope of this PR.

File: src/chart/generateCategoricalChart.tsx (hunk 6 above)
Comment (julianna-langston) [reply src/chart/generateCategoricalChart.tsx]: ditto

File: src/chart/generateCategoricalChart.tsx (hunk 1 above)
Comment (julianna-langston) [reply src/chart/generateCategoricalChart.tsx:1115]: To protect against runtime errors from consumers who are using .jsx, not .tsx. I can imagine someone passing a `null` or something in the prop, and I don't want this method to choke and kill the whole chart.

File: src/chart/generateCategoricalChart.tsx (hunk 7 above)
Comment (julianna-langston) [reply src/chart/generateCategoricalChart.tsx:1277]: It was to help manage if the tooltip should be updated when the image gets resize. I'll rewrite it for clarity.

File: src/chart/generateCategoricalChart.tsx (hunk 8 above)
Comment (julianna-langston) [reply src/chart/generateCategoricalChart.tsx:842]: I scoped the property to the chart, since it's the chart that knows where the tooltip is, where the tickmarks are, what index the tooltip should be on, etc. Do you feel strongly that the property should live on the Tooltip?

Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:842]: I have the same question - as a user I would expect to set a default tooltip index on the tooltip rather than the chart rather than the high level chart. The technical places where things live we can figure out, but the API should align with where someone would expect to see it

File: src/chart/generateCategoricalChart.tsx (hunk 1 above)
Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:1115]: JS users still see types and typehints, imo if they specify something that should be a number as a string that is on them for the most part. But all good to leave the checks

File: src/chart/generateCategoricalChart.tsx (hunk 7 above)
Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:1273]: It looks suspicious because it switched the order of overrides. "Keep" and "updates" are now in the opposite order of what the object destructuring does. Are we 100% certain there are no unintended side effects?

File: src/chart/generateCategoricalChart.tsx (hunk 8 above)
Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:842]: Yes I'm thinking the same. From outside point of view, as a developer using Recharts, I have no knowledge of interest of the inner structure. If it's a property that controls where the Tooltip shows then that's a prop of Tooltip.

Or perhaps there are other use cases for this property that I am missing?

This decision is almost impossible to undo later so we need to get it right the first time

**Comment by julianna-langston** [comment]
How do people feel about these 2 approaches to `defaultIndex` v `activeIndex`:

1. Have only `activeIndex`
Consumers use `activeIndex` to fully manage where the cursor appears. The cursor will _only_ appear at the `activeIndex` position, and if consumers want the tooltip to move with a user's mouse, they'll have to write something to manage updating the index, presumably with the `onMouseMove`. (Will need to figure out a way to tie that in with the accessibility manager.)

2. Separate external (`defaultIndex`) and internal (`activeIndex`)
When consumers set `defaultIndex`, the tooltip will appear at that position at first. Users can interact with the chart, and the tooltip will follow the user. Consumers will not be managing the tooltip position, so the tooltip may wind up moving to another index than the `defaultIndex`.

Note that 2 is what I implemented in this PR so far.

File: src/chart/generateCategoricalChart.tsx (hunk 8 above)
Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:842]: Agreed, especially on the hard-to-undo part. Lets think this through..

Say we look here https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/docs/api-chart-barchart--docs

The chart itself already has:
* `defaultShowTooltip` - this should be on Tooltip too TBH. Also how does this new prop interact with it?

On [Tooltip](https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/docs/api-component-tooltip--docs) we have:
* `active` - this sets the the tooltip to visible always, how does this interact with the new prop? How does it interact with defaultShowTooltip? This is already confusing I think...

Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:842]: From what I can tell `defaultShowTooltip` does't work...

**Comment by ckifer** [comment]
Lets do some pros and cons.

Con of both of these is that they rely on indices, which is something we'll need to move away from eventually. See https://github.com/recharts/recharts/issues/3797 and related issues

> Have only activeIndex

pros:

* gives the user a lot of customization power
* they can pair this with `active` on the tooltip to display the tooltip on first render / always if they want essentially giving the same functionality as 2

cons: 

* requires more management / customization to get it working (could also be a good thing)
* relies on events, our events aren't very optimized and cause a lot of renders


> Separate external (defaultIndex) and internal (activeIndex)


pros:

* gives really nice default functionality
* no need to manage index state
* less need for devs to give examples of how to accomplish this as its done for you
* we can control events internally

cons: 

* semi one-way door decision, if we add this we need to support it
* less customizable potentially
* potential for some undefined behavior with other prop conflicts


questions:

* what happens if you change the defaultIndex with state?

**Comment by julianna-langston** [comment]
@PavelVanecek @ckifer 
This PR is ready for re-review.

**Current functionality**
Consumers can use `<Tooltip defaultIndex={8} />` to set the index they'd like the tooltip to start at.

**Notes**
I kept `defaultIndex` and `activeIndex` separate, because I have a slight preference for it. I think that combining the 2 causes a more onerous experience for developers (see previous comments). That said, if people feel strongly that they should be combined, I'm happy to do it. Just let me know.

I updated the API documentation, and I added test cases. I have full code coverage for every line I changed.

**Comment by bcentdev** [comment]
I need exactly this! Good job! :)

File: src/chart/generateCategoricalChart.tsx [hunk 9]
```diff
@@ -1103,6 +1103,70 @@ export const generateCategoricalChart = ({
...
+          }
+        : {
+            y: independentAxisCoord,
+            x: dependentAxisCoord,
+          };
+
+      // Unlike other chart types, scatter plot's tooltip positions rely on both X and Y coordinates. Only the scatter plot
+      // element knows its own Y coordinates.
+      // If there's a scatter plot, we'll want to grab that element for an interrogation.
+      const scatterPlotElement = this.state.formattedGraphicalItems.find(
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1145]: I wish we didn't have to do special handling for this, but understand why. Just letting it be known

File: src/chart/generateCategoricalChart.tsx [hunk 10]
```diff
@@ -1256,8 +1320,17 @@ export const generateCategoricalChart = ({
...
     };
 
-    // eslint-disable-next-line @typescript-eslint/no-empty-function
-    componentDidUpdate() {}
+    componentDidUpdate(prevProps: CategoricalChartProps) {
+      // Check to see if the Tooltip updated. If so, re-check default tooltip position
+      if (
+        !isChildrenEqual(
+          [findChildByType(prevProps.children, Tooltip)],
+          [findChildByType(this.props.children, Tooltip)],
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1328]: yikes, I wish we didn't have to do this either 😅

**Review by ckifer — APPROVED** [review]
Looks good! Voiced some thoughts but totally understand why things need to be the way they are

**Comment by ckifer** [comment]
🚢
